modules/data.land/R/soil_params_ensemble.R

```diff
@@ -0,0 +1,224 @@
+      sim_q50 <-apply(samples, 2, quantile, probs = 0.50, na.rm = TRUE)
+      sim_q95 <-apply(samples, 2, quantile, probs = 0.95, na.rm = TRUE)
+      quantile_diff <-sum((sim_q5 - q5) ^ 2 + (sim_q50 - q50) ^ 2 + (sim_q95 - q95) ^ 2)
+      return(quantile_diff)
+    }
+    
+    # Optimize alpha0
+    result <- optim(
+      par = 1, # Initial guess for alpha0
+      fn = objective_function,
+      method = "L-BFGS-B",
+      lower = 0.01  # alpha0 must be positive
+    )
+    return(result$par)
+  }
+ 
+  alpha0 <- estimate_alpha0(means, quantiles)
+  if (alpha0 <= 0) {
+    stop("Estimated alpha0 is non-positive, which is invalid.")
+  }
+  alphas <- means * alpha0
+  return(alphas)
+}
+
+
+
+##' A function to estimate the soil parameters based on soil texture data and write the parameter paths into settings
+##' 
+##' @param settings A multi-site settings
+##' @param sand A data frame containing sand fraction in percentage from SoilGrids250m v2.0 with columns "Depth", "Quantile", "Siteid", and "Value"
+##' @param clay A data frame containing clay fraction in percentage from SoilGrids250m v2.0 with columns "Depth", "Quantile", "Siteid", and "Value"
+##' @param silt A data frame containing silt fraction in percentage from SoilGrids250m v2.0 with columns "Depth", "Quantile", "Siteid", and "Value"
+##' @param outdir Provide the path to store the parameter files
+##' @param write_into_settings Whether to write the path of parameter file into the setting. The default is TRUE
+##' 
+##' @example
+##' \dontrun{
+##' 
+##' outdir <- "/projectnb/dietzelab/Cherry/SoilGrids_texture/39NEON"
+##' sand <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/sand_percent.rds") #sand fraction in percentage
+##' clay <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/clay_percent.rds") #clay fraction in percentage
+##' silt <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/silt_percent.rds") #silt fraction in percentage
+##' settings <-read.settings("/projectnb/dietzelab/Cherry/xml/pecan_monthly_SDA_soilwater.xml")
+##' soil_params_ensemble(settings,sand,clay,silt,outdir)
+##'  }
+##'  
+##'  @return Ensemble soil parameter files defined in outdir and file paths in xml file
+##'  @export 
+##'  @author Qianyu Li
+##'  @importFrom magrittr %>%
+##'  
+
+soil_params_ensemble <- function(settings,sand,clay,silt,outdir,write_into_settings=TRUE){
+  
+  # A function to rescale the sums of mean texture fractions to 1 as the original sums are slightly different from 1 for some layers
+  rescale_sum_to_one <- function(sand, clay, silt) {
+    total <- sand + clay + silt
+    rescaled_sand <- sand / total
+    rescaled_clay <- clay / total
+    rescaled_silt <- silt / total
+    return(list(
+      sand = rescaled_sand,
+      clay = rescaled_clay,
+      silt = rescaled_silt))
+  }
+  
+  # A function to reformat the nested list as inputs to "soil2netcdf" function
+  reformat_soil_list <- function(samples_all_depth) {
```

Comment:
Nit: these helper functions are long enough it's hard to scan for where their definitions stop and the parent function resumes. I'd find it more readable if they were defined at the bottom of this file outside of soil_params_ensemble().

Reply:
Fixed.